performance/swiftinterop/src/main/kotlin/org/jetbrains/model/CityMap.kt

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.model
+
+import org.jetbrains.multigraph.*
+
+enum class Transport { CAR, UNDERGROUND, BUS, TROLLEYBUS, TRAM, TAXI, FOOT }
+enum class Interest { SIGHT, CULTURE, PARK, ENTERTAINMENT }
+
+class PlaceAbsenceException(message: String): Exception(message) {}
+
+data class RouteCost(val moneyCost: Double, val timeCost: Double, val interests: Set<Interest>, val transport: Set<Transport>): Cost {
+    override operator fun plus(other: Cost) =
+            if (other is RouteCost) {
+                RouteCost(moneyCost + other.moneyCost, timeCost + other.timeCost,
+                        interests.union(other.interests), transport.union(other.transport))
+            } else {
+                error("Expected type is RouteCost")
+            }
+
+    override operator fun minus(other: Cost) =
+            if (other is RouteCost) {
+                RouteCost(if (moneyCost > other.moneyCost) moneyCost - other.moneyCost else 0.0,
+                        if (timeCost > other.timeCost) timeCost - other.timeCost else 0.0,
+                        interests.subtract(other.interests), transport.subtract(other.transport))
+            } else {
+                error("Expected type is RouteCost")
+            }
+
+    override operator fun compareTo(other: Cost) =
+            if (other is RouteCost) {
+                when {
+                    this == other -> 0
+                    moneyCost < other.moneyCost -> -1
+                    timeCost < other.timeCost -> -1
+                    interests.size < other.interests.size -> -1
+                    transport.size < other.transport.size -> -1
+                    else -> 1
+                }
+            } else {
+                error("Expected type is RouteCost")
+            }
+}
+
+class Place(val geoCoordinateX: Double, val geoCoordinateY: Double, val name: String, val interestCategory: Interest) {
+    companion object {
+        var counter = 0u
+    }
+    val id: UInt
+    init {
+        id = counter
+        counter++
+    }
+
+    override operator fun equals(other: Any?): Boolean {
+        return if (other is Place) geoCoordinateX == other.geoCoordinateX && geoCoordinateY == other.geoCoordinateY &&
```

Comment:
How would this be different from default `equals` implementation if `Place` was a `data class`?

Reply:
Place isn't data class, it has auto-generated id field.

Reply:
> it has auto-generated id field.

How does this prevent `Place` from being a `data class`?

Reply:
Sorry, I was sure that all fields should be in constructor.